x-pack/plugin/deprecation/qa/rest/src/main/java/org/elasticsearch/xpack/deprecation/TestDeprecationHeaderRestAction.java

```diff
@@ -77,19 +75,14 @@ public TestDeprecationHeaderRestAction(Settings settings) {
         this.settings = settings;
     }
 
-    @Override
-    public List<DeprecatedRoute> deprecatedRoutes() {
-        return singletonList(new DeprecatedRoute(GET, "/_test_cluster/deprecated_settings", DEPRECATED_ENDPOINT));
-    }
-
     @Override
     public String getName() {
         return "test_deprecation_header_action";
     }
 
     @Override
     public List<Route> routes() {
-        return Collections.emptyList();
+        return List.of(new Route(GET, "/_test_cluster/deprecated_settings").deprecated(DEPRECATED_ENDPOINT));
```

Comment:
I prefer a builder pattern over a method that returns a new instance. Just reading this line of code makes me think that the `Route` object is mutable when it is not and instead the method returns a new instance. With a builder pattern, I find the code has more clarity at the expense of conciseness.

```java
Route.builder(GET, "/_test_cluster/deprecated_settings").setDeprecationMessage(DEPRECATED_ENDPOINT).build();
```

Static methods for construction similar to the JDK style `of` syntax may provide an alternative that is middle ground for the two?

```java
// normal
Route.of(GET, "/_test_cluster/deprecated_settings");
// a deprecated route
Route.deprecated(GET, "/_test_cluster/deprecated_settings", DEPRECATED_ENDPOINT);
// a replaced route
Route.replaced(GET, "/_test_cluster/deprecated_settings", GET, "/_test_cluster/deprecated_settings2");
// a compatible route - V7 only
Route.compatibleWith(GET, "/_test_cluster/deprecated_settings", CompatibleVersion.V7);
// a compatible route range - works with both 7 and 8
Route.compatibleWith(GET, "/_test_cluster/deprecated_settings", new CompatibleRange(CompatibleVersion.V7, CompatibleVersion.V8));
```

Maybe this makes sense or sparks another idea that's better :)

Reply:
+1, I'll put this in my pipe and smoke it. Thanks for taking a look!

Reply:
I ended up something pretty near what you'd suggested, but a little different. It's got a `Route.of` for the trivial case and a builder for some of the more interesting stuff. Nothing set in stone of course, as you'll be able to see from my string of "ooh, wait, what if we did it like *this*!" commits. 😄